### PR TITLE
Uploader failure handling improvements - DeadLetterQueueHandler implementation, logging improvements

### PR DESCRIPTION
**Pull request by jeffxiang**
This PR simplifies the manual work needed to recover failed uploads in case of issues such as transient S3 slowness / failures. The PR includes the following major changes:
1. Implement `DeadLetterQueueHandler` abstract class: this allows for pluggable implementations of handlers for processing upload tasks that failed after all retries are exhausted. A simple example is `LocalFileDeadLetterQueueHandler` which writes the failures to a local file.
2. Swap out synchronous `S3Client` in `MultiThreadedS3FileUploader` for an async `S3AsyncClient`, so that the `putObject()` call directly returns a future, and aborts the upload if timeout occurs. This fixes the issue where a timeout may not indicate an actual failed upload, which happens when the request is still ongoing but the wrapped future is timed out.
3. Logging improvements: have a clear log message that indicates when all retries have exhausted, to further clarify when a file actually missed upload.

Tested the functionality in unit tests and in real Kafka clusters.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -15,7 +15,7 @@
     </modules>
     <properties>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-        <maven.surefire.version>3.5.1</maven.surefire.version>
+        <maven.surefire.version>3.0.0-M5</maven.surefire.version>
```
Comment (vahidhashemian) [inline pom.xml:18]: Why downgrade this dependency?

**Review by vahidhashemian — APPROVED** [review]

File: pom.xml (hunk 1 above)
Comment (jeffxiang) [reply pom.xml:18]: it's causing issues with maven versions below 3.6.5, we are mostly using 3.6.0.

File: ts-segment-uploader/src/main/java/com/pinterest/kafka/tieredstorage/uploader/MultiThreadedS3FileUploader.java [hunk 2]
```diff
@@ -97,7 +112,7 @@ private int getErrorCode(Throwable throwable, PutObjectResponse putObjectRespons
         if (throwable == null) {
             return putObjectResponse == null ? UPLOAD_GENERAL_ERROR_CODE : putObjectResponse.sdkHttpResponse().statusCode();
         }
-        if (throwable instanceof TimeoutException) {
+        if (throwable instanceof ApiCallTimeoutException || throwable instanceof TimeoutException) {
```
Comment (yisheng-zhou) [inline ts-segment-uploader/src/main/java/com/pinterest/kafka/tieredstorage/uploader/MultiThreadedS3FileUploader.java:115]: Just curious: TimeoutException is from java.util.concurrent while ApiCallTimeoutException is from software.amazon.awssdk.core.exception. Can both of the exceptions appear here?

File: ts-segment-uploader/src/main/java/com/pinterest/kafka/tieredstorage/uploader/DirectoryTreeWatcher.java [hunk 3]
```diff
@@ -315,21 +319,50 @@ private void handleUploadException(UploadTask uploadTask, Throwable throwable, T
...
+        MetricRegistryManager.getInstance(config.getMetricsConfiguration()).incrementCounter(
+                topicPartition.topic(),
+                topicPartition.partition(),
+                UploaderMetrics.UPLOAD_ERROR_METRIC,
+                "exception=" + throwable.getClass().getName(),
+                "cluster=" + environmentProvider.clusterId(),
+                "broker=" + environmentProvider.brokerId(),
+                "offset=" + uploadTask.getOffset()
+        );
+        if (deadLetterQueueHandler != null) {
```
Comment (yisheng-zhou) [inline ts-segment-uploader/src/main/java/com/pinterest/kafka/tieredstorage/uploader/DirectoryTreeWatcher.java:353]: If deadLetterQueueHandler is null, shall we add another logs saying dlq handler is missing so we cannot do anything more than logging?

File: ts-segment-uploader/src/main/java/com/pinterest/kafka/tieredstorage/uploader/DirectoryTreeWatcher.java [hunk 4]
```diff
@@ -315,21 +319,50 @@ private void handleUploadException(UploadTask uploadTask, Throwable throwable, T
...
+            Future<Boolean> result = deadLetterQueueHandler.send(uploadTask, throwable);
+            boolean success;
+            try {
+                success = result.get(deadLetterQueueHandler.getSendTimeoutMs(), TimeUnit.MILLISECONDS);
+            } catch (Exception e) {
+                throw new RuntimeException(String.format("Failed to persist failed upload %s to %s to dead-letter queue." +
+                        " This was a best-effort attempt.", uploadTask.getAbsolutePath(), uploadTask.getUploadDestinationPathString()), e);
```
Comment (yisheng-zhou) [inline ts-segment-uploader/src/main/java/com/pinterest/kafka/tieredstorage/uploader/DirectoryTreeWatcher.java:360]: If DLQ call fails, I dont think we need to fail the main process. The dlq process is apart from main logic. An error log with a lot of details should be good enough.

File: ts-segment-uploader/src/main/java/com/pinterest/kafka/tieredstorage/uploader/dlq/DeadLetterQueueHandler.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+/**
+ * Abstract class for handling failed uploads after retry exhaustion by sending them to a dead letter queue.
+ *
+ * The handler is responsible for sending failed uploads to the dead letter queue and for polling the dead letter queue
+ * for failed uploads to retry. The handler can be specified in the uploader configuration
+ * (see {@link #HANDLER_CLASS_CONFIG_KEY}). The configuration expects the FQDN of the handler class.
+ *
+ * Note that the handler must have a public constructor so that it can be instantiated by reflection.
+ */
+public abstract class DeadLetterQueueHandler {
```
Comment (yisheng-zhou) [inline ts-segment-uploader/src/main/java/com/pinterest/kafka/tieredstorage/uploader/dlq/DeadLetterQueueHandler.java:23]: I like the abstact design. Beside local file dlq and s3 dlq, I like a kafka one the most.

**Pull request merged**